File: Blind150/6.LinkedList/GO/add_two_numbers.go
// https://leetcode.com/problems/add-two-numbers/description/
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {

	add := &ListNode{}

	ans := add

	carry, value := 0, 0

	for l1 != nil && l2 != nil {

		a, b := l1.Val, l2.Val

		carry, value = divmod(a+b+carry, 10)

		add.Next = &ListNode{Val: value}

		l1, l2, add = l1.Next, l2.Next, add.Next

	}

	for l1 != nil {

		carry, value = divmod(l1.Val+carry, 10)

		add.Next = &ListNode{Val: value}

		l1, add = l1.Next, add.Next

	}

	for l2 != nil {

		carry, value = divmod(l2.Val+carry, 10)

		add.Next = &ListNode{Val: value}

		l2, add = l2.Next, add.Next

	}

	if carry != 0 {

		add.Next = &ListNode{Val: carry}
	}

	return ans.Next

}

func divmod(a, b int) (int, int) {
	quotient := a / b
	remainder := a % b
	return quotient, remainder
}

File: Blind150/6.LinkedList/Python/add_two_numbers.py
# https://leetcode.com/problems/add-two-numbers/description/
# Definition for singly-linked list.
from typing import Optional
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:

        add = ListNode()

        ans = add


        carry = 0

        while l1 and l2:

            a,b = l1.val, l2.val

            carry , value = divmod(a+b+carry,10)

            add.next = ListNode(value)

            l1 ,l2,add = l1.next, l2.next, add.next

        while l1:

            carry , value = divmod(l1.val+carry,10)

            add.next = ListNode(value)

            l1,add = l1.next,add.next

        while l2:

            carry , value = divmod(l2.val+carry,10)

            add.next = ListNode(value)

            l2,add = l2.next, add.next

        if carry:

            add.next = ListNode(carry)

        return ans.next
